magenta/models/rl_tuner/rl_tuner.py

```diff
@@ -0,0 +1,2052 @@
+    lengths = np.full(self.q_network.batch_size, 1, dtype=int)
+
+    (action, action_softmax, self.q_network.state_value, 
+    reward_scores, self.reward_rnn.state_value) = self.session.run(
+      [self.predicted_actions, self.action_softmax,
+       self.q_network.state_tensor, self.reward_scores, self.reward_rnn.state_tensor],
+      {self.q_network.melody_sequence: input_batch,
+       self.q_network.initial_state: self.q_network.state_value,
+       self.q_network.lengths: lengths,
+       self.reward_rnn.melody_sequence: input_batch,
+       self.reward_rnn.initial_state: self.reward_rnn.state_value,
+       self.reward_rnn.lengths: lengths})
+
+    # this is apparently not needed
+    #if self.algorithm == 'psi':
+    #  action_scores = np.exp(action_scores)
+
+    reward_scores = np.reshape(reward_scores, (self.num_actions))
+    action_softmax = np.reshape(action_softmax, (self.num_actions))
+    action = np.reshape(action, (self.num_actions))
+
+    if enable_random and random.random() < exploration_p:
+      note = self.get_random_note()
+      if sample_next_obs:
+        return note, note, reward_scores
+      else:
+        return note, reward_scores
+    else:
+      if not sample_next_obs:
+        return action, reward_scores
+      else:
+        obs_note = rl_tuner_ops.sample_softmax(action_softmax)
+        next_obs = np.array(rl_tuner_ops.make_onehot([obs_note],
+                                                   self.num_actions)).flatten()
+        return action, next_obs, reward_scores
+
+  def store(self, observation, state, action, reward, newobservation, newstate, 
+            new_reward_state):
+    """Stores an experience in the model's experience replay buffer.
+
+    One experience consists of an initial observation and internal LSTM state,
+    which led to the execution of an action, the receipt of a reward, and
+    finally a new observation and a new LSTM internal state.
+
+    Args:
+      observation: A one hot encoding of an observed note.
+      state: The internal state of the q_network MelodyRNN LSTM model.
+      action: A one hot encoding of action taken by network.
+      reward: Reward received for taking the action.
+      newobservation: The next observation that resulted from the action.
+        Unless stochastic_observations is True, the action and new
+        observation will be the same.
+      newstate: The internal state of the q_network MelodyRNN that is
+        observed after taking the action.
+      new_reward_state: The internal state of the reward_rnn network that is 
+        observed after taking the action
+    """
+    if self.num_times_store_called % self.dqn_hparams.store_every_nth == 0:
+      self.experience.append((observation, state, action, reward,
+                              newobservation, newstate, new_reward_state))
+      if len(self.experience) > self.dqn_hparams.max_experience:
+        self.experience.popleft()
```

Comment:
rather than doing this, it's more efficient to just instantiate self.experience with a limit:

```
self.experience = deque(maxlen=self.dqn_hparams.max_experience)
```

https://docs.python.org/2/library/collections.html#collections.deque